ludwig/models/modules/sequence_decoders.py

```diff
@@ -15,134 +15,413 @@
+        )
+
+        inputs = first_inputs
+        state = first_state
+
+        # create empty prediction tensor
+        predictions = tf.convert_to_tensor(
+            np.array([]).reshape([batch_size, 0]),
+            dtype=tf.int32   # todo tf2 need to change to tf.int64
+        )
+
+        # beam search
+        for j in range(maximum_iterations):
+            outputs, next_state, next_inputs, finished = decoder.step(
+                j, inputs, state, training=training)
+            inputs = next_inputs
+            state = next_state
+            one_predicted_token = tf.expand_dims(outputs.predicted_ids[:,0], axis=1)
+            predictions = tf.concat([predictions, one_predicted_token], axis=1)
+
+        return predictions
+
+
+    def decoder_inference(
+        self,
+        encoder_output,
+        encoder_end_state=None,
+        training=None
+    ):
+
+        # ================ Setup ================
+        GO_SYMBOL = self.vocab_size
+        END_SYMBOL = 0
+        batch_size = encoder_output.shape[0]
+        encoder_sequence_length = sequence_length_3D(encoder_output)
+
+        # ================ predictions =================
+        greedy_sampler = tfa.seq2seq.GreedyEmbeddingSampler()
+
+        decoder_input = tf.expand_dims(
+            [GO_SYMBOL] * batch_size, 1)
+        start_tokens = tf.fill([batch_size], GO_SYMBOL)
+        end_token = END_SYMBOL
+
+        decoder_emb_inp = self.decoder_embedding(decoder_input)
+
+        if self.attention_mechanism is not None:
+            self.attention_mechanism.setup_memory(
+                encoder_output,
+                memory_sequence_length=encoder_sequence_length
+            )
+
+        decoder_initial_state = self.build_decoder_initial_state(
+            batch_size,
+            encoder_state=encoder_end_state,
+            dtype=tf.float32
+        )
+
+        decoder = tfa.seq2seq.BasicDecoder(
+            cell=self.decoder_rnncell,
+            sampler=greedy_sampler,
+            output_layer=self.dense_layer
+        )
+
+        # ================generate logits ==================
+        maximum_iterations = self.max_sequence_length
+
+        # initialize inference decoder
+        decoder_embedding_matrix = self.decoder_embedding.variables[0]
+        (
+            first_finished,
+            first_inputs,
+            first_state
+        ) = decoder.initialize(
+            decoder_embedding_matrix,
+            start_tokens=start_tokens,
+            end_token=end_token,
+            initial_state=decoder_initial_state
+        )
+
+        inputs = first_inputs
+        state = first_state
+
+        #create empty logits tensor
+        logits = tf.convert_to_tensor(
+            np.array([]).reshape([batch_size, 0, self.num_classes]),
+            dtype=tf.float32
         )
 
-        probabilities_target_sequence = tf.nn.softmax(eval_logits)
+        # build up logits
+        for j in range(maximum_iterations):
```

Comment:
we are not using finished as a stop criterion, we are also generating until the max, we should stop early if possible.